### DISCUSSION TITLE
Is it possible to limit query execution time?

### DISCUSSION THREAD

**Discussion by isidentical** [geldata/gel#1831] [Q&A] [answered]
I have seen queries which took more than an hour (which is actually OK, since I have over 10 million nodes and my query builder is not the most optimal one), but what I wonder is that, is it possible to (either gracefully or forcefully) stop those queries after a certain time (120 seconds)?

As far as I see, when the web server's (gunicorn) workers gets timeouted, the queries still get listed in the process map, drawing about %100 of a single CPU with the name `edgedb {DB_NAME} [local] SELECT`. It would be nice to prevent this without having to `systemctl restart edgedb-server@default` etc.

**Comment by 1st1** [comment, marked as answer]
Query cancellation & timeouts are not yet implemented but is a 1.0 planned feature. You should be able to impose a timeout via nodejs/python APIs though and just close a connection if it hits a timeout.

**Reply by isidentical** [reply]
> Query cancellation & timeouts are not yet implemented but is a 1.0 planned feature.

I see, awesome! 

> You should be able to impose a timeout via nodejs/python APIs though and just close a connection if it hits a timeout.

The `.query()` is blocking, and as far as I can see, when gunicorn tries to timeout them the queries kept running. Will try to investigate the reasoning behind of it, thanks!

**Reply by 1st1** [reply]
Perhaps you can use the asyncio API, where cancellation & timeouts can be done via the framework-level primitives (e.g. `asyncio.wait_for(...)`)

**Reply by isidentical** [reply]
> Perhaps you can use the asyncio API, where cancellation & timeouts can be done via the framework-level primitives (e.g. asyncio.wait_for(...))

Thought that, and probably will use it after my investigation; since it would mean that I had to shift to an async web framework to be consistent in general. Thanks!

**Reply by 1st1** [reply]
Here's a discussion I started on this topic: https://github.com/edgedb/edgedb/discussions/1852

**Reply by isidentical** [reply]
Looks pretty cool, thanks for proposing such an enhancement!
